reconcile/slack_usergroups.py

```diff
@@ -0,0 +1,193 @@
+                continue
+
+            workspace = p['workspace']
+            managed_usergroups = workspace['managedUsergroups']
+            if managed_usergroups is None:
+                continue
+
+            workspace_name = workspace['name']
+            usergroup = p['handle']
+            if usergroup not in managed_usergroups:
+                logging.warning(
+                    '[{}] usergroup {} not in managed usergroups {}'.format(
+                        workspace_name,
+                        usergroup,
+                        managed_usergroups
+                    ))
+                continue
+
+            slack = slack_map[workspace_name]['slack']
+            ugid = slack.get_usergroup_id(usergroup)
+            users_names = [u['slack_username'] for u in r['users']]
+            users = slack.get_users_by_names(users_names)
+
+            channel_names = [] if p['channels'] is None else p['channels']
+            channels = slack.get_channels_by_names(channel_names)
+
+            desired_state.append({
+                "workspace": workspace_name,
+                "usergroup": usergroup,
+                "usergroup_id": ugid,
+                "users": users,
+                "channels": channels,
+            })
+
+    return desired_state
+
+
+def print_diff(current_state, desired_state):
+    for d_state in desired_state:
+        workspace = d_state['workspace']
+        usergroup = d_state['usergroup']
+        c_state = [c for c in current_state
+                   if c['workspace'] == workspace
+                   and c['usergroup'] == usergroup]
+        # at this point we have a single current state item
+        # thanks to the logic in get_desired_state
+        [c_state] = c_state
+
+        channels_to_add = subtract_state(d_state, c_state, 'channels')
+        for c in channels_to_add:
+            logging.info(['add_channel_to_usergroup',
+                          workspace, usergroup, c])
+
+        channels_to_del = subtract_state(c_state, d_state, 'channels')
+        for c in channels_to_del:
+            logging.info(['del_channel_from_usergroup',
+                          workspace, usergroup, c])
+
+        users_to_add = subtract_state(d_state, c_state, 'users')
+        for u in users_to_add:
+            logging.info(['add_user_to_usergroup',
+                          workspace, usergroup, u])
+
+        users_to_del = subtract_state(c_state, d_state, 'users')
+        for u in users_to_del:
+            logging.info(['del_user_from_usergroup',
+                          workspace, usergroup, u])
+
+
+def subtract_state(from_state, subtract_state, type):
+    f = from_state[type]
+    s = subtract_state[type]
+    return [v for k, v in f.items() if k not in s.keys()]
+
+
+def act(desired_state, slack_map):
+    for state in desired_state:
+        workspace = state['workspace']
+        ugid = state['usergroup_id']
+        users = ','.join(state['users'].keys())
```

Comment:
can you move this to inside the update_usergroup_users method? I think it's cleaner... passing an array of users sounds much better than a preformatted string.

Reply:
makes sense